--- firejail-0.9.38.orig/src/firejail/fs_var.c	2016-02-05 04:52:41.000000000 -0800
+++ firejail-0.9.38/src/firejail/fs_var.c	2016-04-29 14:28:49.177776752 -0700
@@ -108,6 +108,9 @@
 }
 	
 void fs_var_log(void) {
+    
+    if(!arg_default_var_log)
+    {
 	build_list("/var/log");
 	
 	// create /var/log if it doesn't exit
@@ -151,6 +154,7 @@
 	}
 	else
 		fprintf(stderr, "Warning: cannot mount tmpfs on top of /var/log\n");
+    }
 }
 
 void fs_var_lib(void) {
