diff -uNPr puppet-4.5.2/puppet.gemspec puppet-4.5.2-modified/puppet.gemspec
--- puppet-4.5.2/puppet.gemspec	1969-12-31 16:00:00.000000000 -0800
+++ puppet-4.5.2-modified/puppet.gemspec	2016-06-29 23:33:00.992701722 -0700
@@ -0,0 +1,17 @@
+# -*- encoding: utf-8 -*-
+
+Gem::Specification.new do |s|
+  s.name = "puppet"
+  s.version = Puppet::PUPPETVERSION
+  s.required_rubygems_version = Gem::Requirement.new(">= 0")
+  s.authors = ["OpenSwitch"]
+  s.date = "2016-06-29"
+  s.summary = "Open source Puppet is a configuration management system"
+  s.description = s.summary
+  s.bindir = 'bin'
+  s.executables  = ['puppet', 'extlookup2hiera']
+  s.homepage = "https://puppetlabs.com/puppet/puppet-open-source"
+  s.require_paths = ["lib"]
+  s.files = %w(LICENSE README.md Rakefile) + Dir.glob('{lib,spec,bin}/**/*')
+  s.specification_version = 3
+end
