From d958ac960472e41fcb8e65259f239a6a11aff38e Mon Sep 17 00:00:00 2001
From: Diego Dompe <dompe@hp.com>
Date: Wed, 1 Apr 2015 09:19:36 -0600
Subject: [PATCH] Disable git hooks that cause trouble when working in Yocto's
 devtool environment

---
 Makefile.am | 52 ++++++++++++++++++++++++++--------------------------
 1 file changed, 26 insertions(+), 26 deletions(-)

diff --git a/Makefile.am b/Makefile.am
index 1a510e9..6fbfcaa 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -164,21 +164,21 @@ CLEAN_LOCAL += clean-pycov
 # Makefile in datapath/linux, needed to get the list of files to
 # distribute, requires GNU make extensions.
 if GNU_MAKE
-ALL_LOCAL += dist-hook-git
-dist-hook-git: distfiles
-	@if test -e $(srcdir)/.git && (git --version) >/dev/null 2>&1; then \
-	  (cd datapath && $(MAKE) distfiles);				    \
-	  (cat distfiles; sed 's|^|datapath/|' datapath/distfiles) |	    \
-	    LC_ALL=C sort -u > all-distfiles;				    \
-	  (cd $(srcdir) && git ls-files) | grep -v '\.gitignore$$' |	    \
-	    LC_ALL=C sort -u > all-gitfiles;				    \
-	  LC_ALL=C comm -1 -3 all-distfiles all-gitfiles > missing-distfiles; \
-	  if test -s missing-distfiles; then				    \
-	    echo "The distribution is missing the following files:";	    \
-	    cat missing-distfiles;					    \
-	    exit 1;							    \
-	  fi;								    \
-	fi
+#ALL_LOCAL += dist-hook-git
+#dist-hook-git: distfiles
+#	@if test -e $(srcdir)/.git && (git --version) >/dev/null 2>&1; then \
+#	  (cd datapath && $(MAKE) distfiles);				    \
+#	  (cat distfiles; sed 's|^|datapath/|' datapath/distfiles) |	    \
+#	    LC_ALL=C sort -u > all-distfiles;				    \
+#	  (cd $(srcdir) && git ls-files) | grep -v '\.gitignore$$' |	    \
+#	    LC_ALL=C sort -u > all-gitfiles;				    \
+#	  LC_ALL=C comm -1 -3 all-distfiles all-gitfiles > missing-distfiles; \
+#	  if test -s missing-distfiles; then				    \
+#	    echo "The distribution is missing the following files:";	    \
+#	    cat missing-distfiles;					    \
+#	    exit 1;							    \
+#	  fi;								    \
+#	fi
 CLEANFILES += all-distfiles all-gitfiles missing-distfiles
 # The following is based on commands for the Automake "distdir" target.
 distfiles: Makefile
@@ -233,17 +233,17 @@ static-check:
 .PHONY: static-check
 
 # Check that assert.h is not used outside a whitelist of files.
-ALL_LOCAL += check-assert-h-usage
-check-assert-h-usage:
-	@if test -e $(srcdir)/.git && (git --version) >/dev/null 2>&1 && \
-	    (cd $(srcdir) && git --no-pager grep -l -E '[<]assert.h[>]') | \
-	    $(EGREP) -v '^lib/(sflow_receiver|vlog).c$$|^tests/'; \
-         then \
-	    echo "Files listed above unexpectedly #include <""assert.h"">."; \
-	    echo "Please use ovs_assert (from util.h) instead of assert."; \
-	    exit 1; \
-	 fi
-.PHONY: check-assert-h-usage
+#ALL_LOCAL += check-assert-h-usage
+#check-assert-h-usage:
+#	@if test -e $(srcdir)/.git && (git --version) >/dev/null 2>&1 && \
+#	    (cd $(srcdir) && git --no-pager grep -l -E '[<]assert.h[>]') | \
+#	    $(EGREP) -v '^lib/(sflow_receiver|vlog).c$$|^tests/'; \
+#         then \
+#	    echo "Files listed above unexpectedly #include <""assert.h"">."; \
+#	    echo "Please use ovs_assert (from util.h) instead of assert."; \
+#	    exit 1; \
+#	 fi
+#.PHONY: check-assert-h-usage
 
 ALL_LOCAL += thread-safety-check
 thread-safety-check:
-- 
1.9.1

