Index: git/LINUX/Makefile
===================================================================
--- git.orig/LINUX/Makefile
+++ git/LINUX/Makefile
@@ -100,6 +100,10 @@ build: $(GET_DRIVERS)
 		O_DRIVERS="$(DRIVERS:%.c=%.o)" modules
 	@ls -l `find . -name \*.ko`
 
+modules_install:
+	$(MAKE) -C $(KSRC) M=$(PWD) CONFIG_NETMAP=m $(MOD_LIST) \
+		EXTRA_CFLAGS='$(EXTRA_CFLAGS)'			\
+		O_DRIVERS="$(DRIVERS:%.c=%.o)" modules_install	
 
 test:
 	@echo "version $(LIN_VER)"
