Index: git/sys/dev/netmap/netmap_mem2.c
===================================================================
--- git.orig/sys/dev/netmap/netmap_mem2.c
+++ git/sys/dev/netmap/netmap_mem2.c
@@ -25,6 +25,7 @@
 
 #ifdef linux
 #include "bsd_glue.h"
+#include <linux/vmalloc.h>
 #endif /* linux */
 
 #ifdef __APPLE__
Index: git/examples/GNUmakefile
===================================================================
--- git.orig/examples/GNUmakefile
+++ git/examples/GNUmakefile
@@ -2,13 +2,13 @@
 # we can just define 'progs' and create custom targets.
 PROGS	=	pkt-gen bridge vale-ctl
 #PROGS += pingd
-PROGS	+= testlock test_select testmmap vale-ctl
+PROGS	+= test_select testmmap vale-ctl
 LIBNETMAP =
 
 CLEANFILES = $(PROGS) *.o
 NO_MAN=
 CFLAGS = -O2 -pipe
-CFLAGS += -Werror -Wall -Wunused-function
+CFLAGS += -Wall -Wunused-function
 CFLAGS += -I ../sys # -I/home/luigi/FreeBSD/head/sys -I../sys
 CFLAGS += -Wextra
 ifdef WITH_PCAP
Index: git/examples/Makefile
===================================================================
--- git.orig/examples/Makefile
+++ git/examples/Makefile
@@ -8,7 +8,7 @@ MORE_PROGS = kern_test
 CLEANFILES = $(PROGS) *.o
 NO_MAN=
 CFLAGS = -O2 -pipe
-CFLAGS += -Werror -Wall -Wunused-function
+CFLAGS += -Wall -Wunused-function
 CFLAGS += -I ../sys # -I/home/luigi/FreeBSD/head/sys -I../sys
 CFLAGS += -Wextra
 .ifdef WITH_PCAP
