Index: linux-3.9.6/drivers/i2c/busses/i2c-ismt.c
===================================================================
--- linux-3.9.6.orig/drivers/i2c/busses/i2c-ismt.c	2014-07-18 18:14:44.931826515 -0500
+++ linux-3.9.6/drivers/i2c/busses/i2c-ismt.c	2014-07-18 14:07:50.647839122 -0500
@@ -71,6 +71,7 @@
 #include <linux/i2c.h>
 #include <linux/acpi.h>
 #include <linux/interrupt.h>
+#include <linux/delay.h>

 #include <asm-generic/io-64-nonatomic-lo-hi.h>

@@ -193,9 +194,12 @@ static const DEFINE_PCI_DEVICE_TABLE(ism
 MODULE_DEVICE_TABLE(pci, ismt_ids);

 /* Bus speed control bits for slow debuggers - refer to the docs for usage */
-static unsigned int bus_speed;
+static unsigned int bus_speed = 100;
+static unsigned int delay = 1000;
 module_param(bus_speed, uint, S_IRUGO);
-MODULE_PARM_DESC(bus_speed, "Bus Speed in kHz (0 = BIOS default)");
+MODULE_PARM_DESC(bus_speed, "Bus Speed in kHz (1000 by default)");
+module_param(delay, uint, S_IRUGO);
+MODULE_PARM_DESC(delay, "Delay in microsecs before access (1000 by default)");


 /**
@@ -344,6 +344,7 @@
 			data->word = dma_buffer[0] | (dma_buffer[1] << 8);
 			break;
 		case I2C_SMBUS_BLOCK_DATA:
+		case I2C_SMBUS_I2C_BLOCK_DATA:	
 			memcpy(&data->block[1], dma_buffer, desc->rxbytes);
 			data->block[0] = desc->rxbytes;
 			break;
@@ -393,6 +397,9 @@ static int ismt_access(struct i2c_adapte
	struct ismt_priv *priv = i2c_get_adapdata(adap);
	struct device *dev = &priv->pci_dev->dev;

+        if (delay > 0)
+            udelay(delay);
+
	desc = &priv->hw[priv->head];

	/* Initialize the descriptor */
@@ -505,7 +506,40 @@
 			desc->control |= (ISMT_DESC_BLK | ISMT_DESC_CWRL);
 		}
 		break;
+	case I2C_SMBUS_I2C_BLOCK_DATA:
+		/* Make sure the lenght is valid */
+		if (data->block[0] < 1)
+			data->block[0] = 1;
 
+		if (data->block[0] > I2C_SMBUS_BLOCK_MAX)
+			data->block[0] = I2C_SMBUS_BLOCK_MAX;
+
+		if (read_write == I2C_SMBUS_WRITE) {
+			/* i2c Block Write */
+			dev_dbg(dev, "I2C_SMBUS_I2C_BLOCK_DATA:  WRITE\n");
+			dma_size = data->block[0] + 1;
+			dma_direction = DMA_TO_DEVICE;
+			desc->wr_len_cmd = dma_size;
+			desc->control |= ISMT_DESC_I2C;
+			priv->dma_buffer[0] = command;
+			memcpy(&priv->dma_buffer[1], &data->block[1], dma_size);
+		} else {
+			/* i2c Block Read */
+			dev_dbg(dev, "I2C_SMBUS_I2C_BLOCK_DATA:  READ\n");
+			dma_size = data->block[0];
+			dma_direction = DMA_FROM_DEVICE;
+			desc->rd_len = dma_size;
+			desc->wr_len_cmd = command;
+			desc->control |= (ISMT_DESC_I2C | ISMT_DESC_CWRL);
+			/*
+			 * Per the "Table 15-15. I2C Commands",
+			 * in the External Design Specification (EDS),
+			 * (Document Number: 508084, Revision: 2.0),
+			 * the _rw bit must be 0
+			 */
+			desc->tgtaddr_rw = ISMT_DESC_ADDR_RW(addr, 0);
+		}
+		break;
 	default:
 		dev_err(dev, "Unsupported transaction %d\n",
 			size);
@@ -579,6 +613,7 @@
 	       I2C_FUNC_SMBUS_WORD_DATA		|
 	       I2C_FUNC_SMBUS_PROC_CALL		|
 	       I2C_FUNC_SMBUS_BLOCK_DATA	|
+	       I2C_FUNC_SMBUS_I2C_BLOCK |
 	       I2C_FUNC_SMBUS_PEC;
 }
