Index: kernel-source/drivers/net/ethernet/intel/igb/igb_main.c
===================================================================
--- kernel-source.orig/drivers/net/ethernet/intel/igb/igb_main.c
+++ kernel-source/drivers/net/ethernet/intel/igb/igb_main.c
@@ -1561,6 +1561,8 @@ void igb_power_up_link(struct igb_adapte
 		igb_power_up_phy_copper(&adapter->hw);
 	else
 		igb_power_up_serdes_link_82575(&adapter->hw);
+
+	igb_setup_link(&adapter->hw);
 }
 
 /**
